### DISCUSSION TITLE
FYI: Subscribing to node-matter

### DISCUSSION THREAD

**Discussion by javierbarellano4v** [mfucci/node-matter#233] [General]
Just a FYI...

With this setup:

node-matter/MacOS
Alexa app/Android OS/tablet
Alexa hub/Echo Dot
chip-tool/MacOS
Node-matter ("light") paired with both Alexa and chip-tool.

I sent a subscribe command on OnOff to node-matter from the chip-tool (interactive mode). 

So, whenever the "light" state was changed (via Alexa app or Echo Dot), the chip-tool would display in its running log.

Here is an excerpt:

**1675965964220] [57919:5515424] CHIP: [TOO] Endpoint: 1 Cluster: 0x0000_0006 Attribute 0x0000_0000 DataVersion: 17
[**1675965964220] [57919:5515424] CHIP: [TOO]   OnOff: TRUE
[1675965964220] [57919:5515424] CHIP: [DMG] Refresh LivenessCheckTime for 26000 milliseconds with SubscriptionId = 0x00000000 Peer = 01:0000000000000001****
[1675965964220] [57919:5515424] CHIP: [EM] Piggybacking Ack for MessageCounter:3688376575 on exchange: 35270r
[1675965964220] [57919:5515424] CHIP: [IN] Prepared secure message 0x7f7e7f80a868 to 0x0000000000000001 (1)  of type 0x1 and protocolId (0, 1) on exchange 35270r with MessageCounter:103011186.
[1675965964220] [57919:5515424] CHIP: [IN] Sending encrypted msg 0x7f7e7f80a868 with MessageCounter:103011186 to 0x0000000000000001 (1) at monotonic time: 0000000037161295 msec
[1675965964221] [57919:5515424] CHIP: [EM] Received message of type 0x10 with protocolId (0, 0) and MessageCounter:3688376576 on exchange 35270r
[1675965964221] [57919:5515424] CHIP: [EM] Found matching exchange: 35270r, Delegate: 0x0
[1675965964221] [57919:5515424] CHIP: [EM] Rxd Ack; Removing MessageCounter:103011186 from Retrans Table on exchange 35270r
[1675965964221] [57919:5515424] CHIP: [EM] Removed CHIP MessageCounter:103011186 from RetransTable on exchange 35270r

**Comment by Apollon77** [comment]
Honestly, no idea what that log want to say ;-)

But Subscription in node-matter is currently implemented in a buggy way. There is a PR already to adress this which needs to be merged, so will be fixed soon and then subscriptions should be at least as the specs require. I would then recheck that once we have corrected the issue

**Comment by javierbarellano4v** [comment]
@Apollon77 My posting was the use case where a "device" is subscribing to an attribute on node-matter.

I'm assuming that the PR (h[ttps://github.com/mfucci/node-matter/pull/219](url)) you mention includes functionality for the flip situation where node-matter subscribes to an attribute on another device (among other things). I hope.  Thanks.

**Reply by Apollon77** [reply]
No, the PR adresses "a controller/other device requests a subscribe on node-matter used as device", so should be your use case. The current implementation in main branch is not following the specs. The PR willmake it compatible ... verified with Google, Apple, chip-tool

**Reply by javierbarellano4v** [reply]
I see. Will it ever support the opposite , “node-matter used as  a device requests a subscribe on other device”? Thanks.

**Reply by Apollon77** [reply]
Thats all questions about usecases. We did not really started looking into that whole group topics, scenes and such. I did not strted looking into the whole think about "Local linking" or such case (e.g. a sensor triggers a function on an other device). My understanding so far was that these are group messages and not really "subscriptions" .... but as said, not looked into it that deep so far.

So the answer to "will it ever" is a clear: If it makes sense then sure :-)
And in fact PRs are always welcome :-)

And BTW: Just because Chip-tool (as a protocol reference implementation) can do something it do not really mean that it makes sense in real world scenarios. ;-) So be careful with assumptions!

**Reply by javierbarellano4v** [reply]
According to the specs, subscription can support the use case subscribing one device/node's (subscriber) to another one's data (publisher), e.g., a  first device measuring temperature and a second device wants to read out or get notified if temperature changed on the first device.

But, as you say PRs are welcome.

**Reply by Apollon77** [reply]
Bzw: do you have the specs location for me where this is described?

**Comment by javierbarellano4v** [comment]
Yes.

8.5.Subscribe Interaction ..........page 420

**Reply by Apollon77** [reply]
Ok, this is the "Normal" controller subscribes with a device interaction ... so multi controller pairing ... This we already support. Then the "device" needs to be paired as controller with node-matter and then it can subscribe.
I have one node-matter here which is paired to Apple, Goole and Alexa at the same time and all can interact with it ... I can try to pair chiltool also tomorrow  ... and as said .. that should already work (ok afetr the above mentioned PR is merged)

**Reply by mfucci** [reply]
My understanding is that the two devices inside the same fabric should be able to communicate with each other, they don't need to be admin or to have provisioned the device.
For instance, I use Google Home to commission device A and device B, now device A can talk / subscribe to device B directly (using CASE).
The code to do that should be working, but there are no easy API on top to make this work (the current hack would be to start a Device and when it is provision, reuse the fabric inside a Controller).

It is on the roadmap to merge Device and Controller into a Node which will support client and server APIs.

**Reply by javierbarellano4v** [reply]
Yes, that was my understanding too. I had not been able to reply.

Just curious, assuming you’re programmatically subscribing to a device?

**Reply by javierbarellano4v** [reply]
@mfucci let me pose my question differently (my earlier question was superfluous - programmatically of course, thinking more at the command line with my question). 

Is the code to subscribe in node-matter available ?  Or are you saying that subscription in general is feasible but not yet available?

**Reply by Apollon77** [reply]
node-matter already have code to subscribe to values and send (when acting as device) or receive (when acting as controller) the data. We optimize this currently in PRs and APIs. So what exactly is your question? ;-)

**Comment by javierbarellano4v** [comment]
I should've been more precise in my question.

Correct,  able to subscribe to data on node-matter using the chip-tool's subscribe command and getting notifications in chip-tool when toggling the power on/off in node-matter via Alexa.

I was just trying to see how to subscribe to another devices' data from node-matter, in general. This is not exposed, e.g., like a subscription command similar to what the chip-tool supports.

So it appears I need to consult the PRs and examine the APIs.

**Reply by Apollon77** [reply]
Ok please provide more details whyt is setup how becasue Is till do not get it (and could also try to reproduce it):
1.) node-matter is running as "Device.ts" right?
2.) You have paired Alexa "as controller "to the node-matter device, right?
3.) You also paired chip-tol "as controller" to the node-matter device (kind of as second controller by getting the pairing code from alexa), right?
4.) Now you subscribe on light status in chip-tool

Is it that way? If not please state how your setup is.

If it is as above then you run into the "issue" that main branch is not answerign corectly to subscribe requests and so most likely chip tool do not subscribe correctly. This will be fixed by the one PR mentioned above - I hope we can merge that ltest on the weekend. Then we could retest all this.

**Reply by javierbarellano4v** [reply]
@Apollon77 Well, I discovered that I did do step 3, I thought that I had not paired chip-tool with node-matter. 

I had been testing multi-admin, etc. and forgot that it was paired to chip-tool! Ugh! :-(

So, I stand corrected. I look forward to the fix.  

I may be misunderstanding what @mfucci posted earlier regarding his success subscribing between devices,.. functionality not merged yet, maybe? 

Thanks.

**Reply by Apollon77** [reply]
The case @mfucci described should work yes, but honestly I never tried something like this. I worked a lot to allow clean subscriptions for controller <-> device ... lets see where we come to once this all is merged and works :-)

**Reply by javierbarellano4v** [reply]
@Apollon77 I'm assuming that PR for this issue hasn't been merged in, right?

**Reply by Apollon77** [reply]
We merged one on the weekend. There are others left that also optimize things but the basic PR is in main branch. Try it

**Comment by javierbarellano4v** [comment]
Great! 

just make sure I get it..., is this ok?

1. Get the sources git clone https://github.com/mfucci/node-matter.git 
2. cd node-matter
3. Install the dependencies: npm i
4. Build: npm run build
5. Run: node build/Device.js
6. (optional) Install globally: npm -g i . and run it with matter

**Reply by Apollon77** [reply]
Basically yes (i would not do it globally ... but up to you)

**Reply by javierbarellano4v** [reply]
@Apollon77 I updated node-matter.

I tried to subscribe from chip-tool to node-matter. (without chip-tool pairing with node-matter) 

onoff subscribe on-off 1 1 1 1 

![image](https://user-images.githubusercontent.com/97479879/220412659-62a04d75-5331-4783-9ac8-31f971b5fbc3.png)

Assumption, chip-tool will find node on the fabric?

**Reply by Apollon77** [reply]
I think you need to pair before because it needs to have a valid session for subscriptions. So that needs to be done before. I will check specs tomorrow

**Reply by javierbarellano4v** [reply]
What's perplexing is when I run "discover commissionables" or "discover commissioners" in the chip-too,  node-matter doesn't show up.

**Reply by javierbarellano4v** [reply]
@Apollon77 @mfucci 

I think the problem or twist is that the chip-tool is essentially a Controller. 

If node-matter has been previously commissioned by a Controller, e.g., Amazon Echo Dot, leads me to some questions:

1. Would chip-tool be on the same fabric as the Echo Dot? Assuming it is.
2. If there are multiple commissioners, is that really a multi-admin scenario? Not sure.
3. With multiple controllers, they're essentially peers in the sense of control, so how does a new controller in the same fabric get commissioned or "onboarded"?

I'm asking some basic questions. that I thought was very clear but I need to confirm?

Maybe I should pull this out to a new thread in the Q&A section?

Thanks.

**Reply by Apollon77** [reply]
After pairing with a controller this controller should have a function to get another pairing code for another controller. When this pairing code is requested a new commissioning window on the device is opened and then this is also discoverable then. A new pairing process creates another fabric on the node.

**Reply by Apollon77** [reply]
Ookk .. I also tested that out and subscriptions do work with chip-tool. YOur error above comes from chip tool not being able to discover the node (whyever). Locally for me I already implemented Persistence, so that worked well because I could simply restart the process and it was working  ... But I still do not get why chiptool then closes again after successfully having the subscription as controller ... maybe I did not found the right parameter to leave it running because so it makes no sense bevause it never gets the subscription status updates :-(

**Reply by javierbarellano4v** [reply]
@Apollon77 Just make sure I understand.

I'm assuming your 'node-matter' device was commissioned/paired with Alexa, Google etc. initially.

Then you attempted to subscribe to node-matter using chip-tool but you DID NOT pair chip-tool with node-matter device beforehand?

Thanks.

**Reply by Apollon77** [reply]
My test was to just pair with chip tool directly. There should be no change from the view of chip-tool. i can also try a multi pairing tomorow
